# Classificação de Doença Hepática com Aprendizagem de Máquina
## Dataset Indian Liver Patient Records (ILPD — UCI)

---
# Fase 1 — Preparação e Entendimento do Dataset

**Objetivo:** Carregar o dataset, compreender sua estrutura e documentar todas as características relevantes antes de iniciar qualquer análise ou modelagem.

## 1.1 — Importações e Carregamento do Dataset

In [1]:
%pip install kagglehub -q

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.1.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import kagglehub
import pandas as pd
import numpy as np
import os

# Download do dataset via Kaggle
path = kagglehub.dataset_download("uciml/indian-liver-patient-records")
print("Path to dataset files:", path)
print("Arquivos disponíveis:", os.listdir(path))

Path to dataset files: C:\Users\lucas\.cache\kagglehub\datasets\uciml\indian-liver-patient-records\versions\1
Arquivos disponíveis: ['indian_liver_patient.csv']


In [3]:
# Carregamento do dataset
df = pd.read_csv(os.path.join(path, "indian_liver_patient.csv"))

print("Dataset carregado com sucesso.")
print(f"Dimensão: {df.shape[0]} registros × {df.shape[1]} colunas")

Dataset carregado com sucesso.
Dimensão: 583 registros × 11 colunas


In [4]:
# Primeiras linhas
df.head()

,Age,Gender,Total_Bilirubin,Direct_Bilirubin,Alkaline_Phosphotase,Alamine_Aminotransferase,Aspartate_Aminotransferase,Total_Protiens,Albumin,Albumin_and_Globulin_Ratio,Dataset
0,65,Female,0.7,0.1,187,16,18,6.8,3.3,0.90,1
1,62,Male,10.9,5.5,699,64,100,7.5,3.2,0.74,1
2,62,Male,7.3,4.1,490,60,68,7.0,3.3,0.89,1
3,58,Male,1.0,0.4,182,14,20,6.8,3.4,1.00,1
4,72,Male,3.9,2.0,195,27,59,7.3,2.4,0.40,1


In [5]:
# Últimas linhas
df.tail()

,Age,Gender,Total_Bilirubin,Direct_Bilirubin,Alkaline_Phosphotase,Alamine_Aminotransferase,Aspartate_Aminotransferase,Total_Protiens,Albumin,Albumin_and_Globulin_Ratio,Dataset
578,60,Male,0.5,0.1,500,20,34,5.9,1.6,0.37,2
579,40,Male,0.6,0.1,98,35,31,6.0,3.2,1.10,1
580,52,Male,0.8,0.2,245,48,49,6.4,3.2,1.00,1
581,31,Male,1.3,0.5,184,29,32,6.8,3.4,1.00,1
582,38,Male,1.0,0.3,216,21,24,7.3,4.4,1.50,2


## 1.2 — Identificação das Colunas

| Coluna | Descrição | Tipo Esperado |
|---|---|---|
| `Age` | Idade do paciente | Numérico (inteiro) |
| `Gender` | Gênero do paciente (Male/Female) | Categórico |
| `Total_Bilirubin` | Bilirrubina total | Numérico (float) |
| `Direct_Bilirubin` | Bilirrubina direta | Numérico (float) |
| `Alkaline_Phosphotase` | Fosfatase alcalina | Numérico (inteiro) |
| `Alamine_Aminotransferase` | SGPT / ALT (alanina aminotransferase) | Numérico (inteiro) |
| `Aspartate_Aminotransferase` | SGOT / AST (aspartato aminotransferase) | Numérico (inteiro) |
| `Total_Protiens` | Proteínas totais | Numérico (float) |
| `Albumin` | Albumina | Numérico (float) |
| `Albumin_and_Globulin_Ratio` | Razão albumina/globulina (A/G) | Numérico (float) |
| `Dataset` | **Variável-alvo** (1 = doença hepática, 2 = sem doença) | Numérico (inteiro) |

In [6]:
# Listagem das colunas
print("Colunas do dataset:")
for i, col in enumerate(df.columns, 1):
    print(f"  {i:2d}. {col}")

Colunas do dataset:
   1. Age
   2. Gender
   3. Total_Bilirubin
   4. Direct_Bilirubin
   5. Alkaline_Phosphotase
   6. Alamine_Aminotransferase
   7. Aspartate_Aminotransferase
   8. Total_Protiens
   9. Albumin
  10. Albumin_and_Globulin_Ratio
  11. Dataset


## 1.3 — Tipos de Dados

In [7]:
# Tipos de dados de cada coluna
print("Tipos de dados:")
print(df.dtypes)
print(f"\nTotal de colunas numéricas: {df.select_dtypes(include='number').shape[1]}")
print(f"Total de colunas categóricas: {df.select_dtypes(include='object').shape[1]}")

Tipos de dados:


Age                             int64
Gender                         object
Total_Bilirubin               float64
Direct_Bilirubin              float64
Alkaline_Phosphotase            int64
Alamine_Aminotransferase        int64
Aspartate_Aminotransferase      int64
Total_Protiens                float64
Albumin                       float64
Albumin_and_Globulin_Ratio    float64
Dataset                         int64
dtype: object

Total de colunas numéricas: 10
Total de colunas categóricas: 1


In [8]:
# Informações gerais do DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 583 entries, 0 to 582
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Age                         583 non-null    int64  
 1   Gender                      583 non-null    object 
 2   Total_Bilirubin             583 non-null    float64
 3   Direct_Bilirubin            583 non-null    float64
 4   Alkaline_Phosphotase        583 non-null    int64  
 5   Alamine_Aminotransferase    583 non-null    int64  
 6   Aspartate_Aminotransferase  583 non-null    int64  
 7   Total_Protiens              583 non-null    float64
 8   Albumin                     583 non-null    float64
 9   Albumin_and_Globulin_Ratio  579 non-null    float64
 10  Dataset                     583 non-null    int64  
dtypes: float64(5), int64(5), object(1)
memory usage: 50.2+ KB


**Observações sobre os tipos de dados:**
- As 9 features numéricas estão com tipos compatíveis (`int64` e `float64`).
- `Gender` é do tipo `object`, o que confirma sua natureza categórica.
- `Dataset` (variável-alvo) é `int64`, contendo os valores 1 e 2.
- `Albumin_and_Globulin_Ratio` é `float64` e possui 579 valores não-nulos (4 ausentes).

## 1.4 — Tamanho do Dataset

In [9]:
n_registros, n_colunas = df.shape
n_features = n_colunas - 1  # excluindo a variável-alvo

print(f"Número de registros: {n_registros}")
print(f"Número de colunas: {n_colunas}")
print(f"Número de features preditoras: {n_features}")
print(f"Variável-alvo: Dataset")
print(f"\nUso de memória: {df.memory_usage(deep=True).sum() / 1024:.1f} KB")

Número de registros: 583
Número de colunas: 11
Número de features preditoras: 10
Variável-alvo: Dataset

Uso de memória: 76.1 KB


## 1.5 — Análise Inicial da Variável-Alvo

In [10]:
# Distribuição da variável-alvo
contagem_alvo = df["Dataset"].value_counts().sort_index()
proporcao_alvo = df["Dataset"].value_counts(normalize=True).sort_index() * 100

print("Distribuição da variável-alvo (Dataset):")
print(f"  Classe 1 (doença hepática):     {contagem_alvo[1]:>4d} registros ({proporcao_alvo[1]:.1f}%)")
print(f"  Classe 2 (sem doença hepática):  {contagem_alvo[2]:>4d} registros ({proporcao_alvo[2]:.1f}%)")
print(f"\n  Razão entre classes (1:2): {contagem_alvo[1] / contagem_alvo[2]:.2f} : 1")
print(f"\n⚠️  O dataset é desbalanceado: a classe majoritária (doença) representa ~{proporcao_alvo[1]:.0f}% dos dados.")

Distribuição da variável-alvo (Dataset):
  Classe 1 (doença hepática):      416 registros (71.4%)
  Classe 2 (sem doença hepática):   167 registros (28.6%)

  Razão entre classes (1:2): 2.49 : 1

⚠️  O dataset é desbalanceado: a classe majoritária (doença) representa ~71% dos dados.


In [11]:
# Verificação de que a variável-alvo contém apenas os valores esperados (1 e 2)
valores_unicos_alvo = sorted(df["Dataset"].unique())
print(f"Valores únicos em 'Dataset': {valores_unicos_alvo}")

assert valores_unicos_alvo == [1, 2], "A variável-alvo contém valores inesperados!"
print("✓ Variável-alvo contém apenas os valores 1 e 2, conforme esperado.")

Valores únicos em 'Dataset': [np.int64(1), np.int64(2)]
✓ Variável-alvo contém apenas os valores 1 e 2, conforme esperado.


## 1.6 — Identificação de Valores Ausentes

In [12]:
# Contagem de valores ausentes por coluna
valores_ausentes = df.isnull().sum()
percentual_ausentes = (df.isnull().sum() / len(df)) * 100

resumo_ausentes = pd.DataFrame({
    "Ausentes": valores_ausentes,
    "Percentual (%)": percentual_ausentes.round(2)
})

print("Valores ausentes por coluna:")
print(resumo_ausentes)
print(f"\nTotal de valores ausentes no dataset: {valores_ausentes.sum()}")

Valores ausentes por coluna:
                            Ausentes  Percentual (%)
Age                                0            0.00
Gender                             0            0.00
Total_Bilirubin                    0            0.00
Direct_Bilirubin                   0            0.00
Alkaline_Phosphotase               0            0.00
Alamine_Aminotransferase           0            0.00
Aspartate_Aminotransferase         0            0.00
Total_Protiens                     0            0.00
Albumin                            0            0.00
Albumin_and_Globulin_Ratio         4            0.69
Dataset                            0            0.00

Total de valores ausentes no dataset: 4


In [13]:
# Detalhamento dos registros com valores ausentes
registros_ausentes = df[df["Albumin_and_Globulin_Ratio"].isnull()]
print(f"Registros com Albumin_and_Globulin_Ratio ausente: {len(registros_ausentes)}")
print()
registros_ausentes

Registros com Albumin_and_Globulin_Ratio ausente: 4



,Age,Gender,Total_Bilirubin,Direct_Bilirubin,Alkaline_Phosphotase,Alamine_Aminotransferase,Aspartate_Aminotransferase,Total_Protiens,Albumin,Albumin_and_Globulin_Ratio,Dataset
209,45,Female,0.9,0.3,189,23,33,6.6,3.9,NaN,1
241,51,Male,0.8,0.2,230,24,46,6.5,3.1,NaN,1
253,35,Female,0.6,0.2,180,12,15,5.2,2.7,NaN,2
312,27,Male,1.3,0.6,106,25,54,8.5,4.8,NaN,2


In [14]:
# Verificar se os valores ausentes estão associados a alguma classe específica
print("Distribuição da variável-alvo nos registros com valores ausentes:")
print(registros_ausentes["Dataset"].value_counts().sort_index())
print()
print("Distribuição da variável-alvo no dataset completo:")
print(df["Dataset"].value_counts().sort_index())

Distribuição da variável-alvo nos registros com valores ausentes:
Dataset
1    2
2    2
Name: count, dtype: int64

Distribuição da variável-alvo no dataset completo:


Dataset
1    416
2    167
Name: count, dtype: int64


**Observações sobre valores ausentes:**
- Apenas a coluna `Albumin_and_Globulin_Ratio` possui valores ausentes (4 registros — 0,69% do dataset).
- A quantidade é pequena e pode ser tratada por imputação na fase de pré-processamento.
- A estratégia de imputação será definida na Fase 3, utilizando somente os dados de treinamento para evitar data leakage.

## 1.7 — Identificação de Duplicatas

In [15]:
# Contagem de registros duplicados
n_duplicatas = df.duplicated().sum()
n_registros_envolvidos = df[df.duplicated(keep=False)].shape[0]

print(f"Registros duplicados (excluindo a primeira ocorrência): {n_duplicatas}")
print(f"Total de registros envolvidos em duplicações: {n_registros_envolvidos}")

Registros duplicados (excluindo a primeira ocorrência): 13
Total de registros envolvidos em duplicações: 26


In [16]:
# Exibir os registros duplicados para análise
duplicatas = df[df.duplicated(keep=False)].sort_values(by=list(df.columns))
duplicatas

,Age,Gender,Total_Bilirubin,Direct_Bilirubin,Alkaline_Phosphotase,Alamine_Aminotransferase,Aspartate_Aminotransferase,Total_Protiens,Albumin,Albumin_and_Globulin_Ratio,Dataset
137,18,Male,0.8,0.2,282,72,140,5.5,2.5,0.80,1
138,18,Male,0.8,0.2,282,72,140,5.5,2.5,0.80,1
142,30,Male,1.6,0.4,332,84,139,5.6,2.7,0.90,1
143,30,Male,1.6,0.4,332,84,139,5.6,2.7,0.90,1
173,31,Male,0.6,0.1,175,48,34,6.0,3.7,1.60,1
174,31,Male,0.6,0.1,175,48,34,6.0,3.7,1.60,1
25,34,Male,4.1,2.0,289,875,731,5.0,2.7,1.10,1
26,34,Male,4.1,2.0,289,875,731,5.0,2.7,1.10,1
107,36,Male,0.8,0.2,158,29,39,6.0,2.2,0.50,2
108,36,Male,0.8,0.2,158,29,39,6.0,2.2,0.50,2


**Observações sobre duplicatas:**
- Existem **13 registros duplicados** (26 registros envolvidos — 13 pares).
- Cada duplicata aparece exatamente 2 vezes.
- Como o dataset contém dados clínicos, é plausível que pacientes diferentes apresentem exatamente os mesmos valores de exames. Entretanto, a duplicação exata em todas as 11 colunas sugere que podem ser registros repetidos por erro de entrada.
- **Decisão:** As duplicatas serão mantidas nesta fase. Se necessário, a remoção será avaliada na fase de pré-processamento com base no impacto na modelagem. A quantidade é pequena (2,2% do dataset) e a remoção não alteraria significativamente a distribuição dos dados.

## 1.8 — Verificação de Inconsistências

In [17]:
# Verificar valores negativos em colunas numéricas
colunas_numericas = df.select_dtypes(include="number").columns
negativos = (df[colunas_numericas] < 0).sum()

print("Valores negativos por coluna numérica:")
print(negativos)

if negativos.sum() == 0:
    print("\n✓ Nenhum valor negativo encontrado. Todas as colunas numéricas possuem valores ≥ 0.")
else:
    print(f"\n⚠️ Foram encontrados {negativos.sum()} valores negativos.")

Valores negativos por coluna numérica:
Age                           0
Total_Bilirubin               0
Direct_Bilirubin              0
Alkaline_Phosphotase          0
Alamine_Aminotransferase      0
Aspartate_Aminotransferase    0
Total_Protiens                0
Albumin                       0
Albumin_and_Globulin_Ratio    0
Dataset                       0
dtype: int64

✓ Nenhum valor negativo encontrado. Todas as colunas numéricas possuem valores ≥ 0.


In [18]:
# Verificar valores únicos da variável Gender
valores_gender = df["Gender"].unique()
print(f"Valores únicos em 'Gender': {valores_gender}")
print(f"Contagem:")
print(df["Gender"].value_counts())

assert set(valores_gender) == {"Male", "Female"}, "Gender contém valores inesperados!"
print("\n✓ Gender contém apenas 'Male' e 'Female', conforme esperado.")

Valores únicos em 'Gender': ['Female' 'Male']
Contagem:
Gender
Male      441
Female    142
Name: count, dtype: int64

✓ Gender contém apenas 'Male' e 'Female', conforme esperado.


In [19]:
# Estatísticas descritivas de todas as colunas numéricas
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Age,583.0,44.746141,16.189833,4.0,33.0,45.00,58.0,90.0
Total_Bilirubin,583.0,3.298799,6.209522,0.4,0.8,1.00,2.6,75.0
Direct_Bilirubin,583.0,1.486106,2.808498,0.1,0.2,0.30,1.3,19.7
Alkaline_Phosphotase,583.0,290.576329,242.937989,63.0,175.5,208.00,298.0,2110.0
Alamine_Aminotransferase,583.0,80.713551,182.620356,10.0,23.0,35.00,60.5,2000.0
Aspartate_Aminotransferase,583.0,109.910806,288.918529,10.0,25.0,42.00,87.0,4929.0
Total_Protiens,583.0,6.483190,1.085451,2.7,5.8,6.60,7.2,9.6
Albumin,583.0,3.141852,0.795519,0.9,2.6,3.10,3.8,5.5
Albumin_and_Globulin_Ratio,579.0,0.947064,0.319592,0.3,0.7,0.93,1.1,2.8
Dataset,583.0,1.286449,0.452490,1.0,1.0,1.00,2.0,2.0


In [20]:
# Verificação de valores plausíveis
print("Verificação de plausibilidade dos valores:")
print(f"  Age:  min = {df['Age'].min()}, max = {df['Age'].max()}")
print(f"    → Intervalo [4, 90] — plausível, inclui pacientes pediátricos e idosos.")
print()
print(f"  Total_Bilirubin: min = {df['Total_Bilirubin'].min()}, max = {df['Total_Bilirubin'].max()}")
print(f"    → Valor máximo de 75.0 é extremamente elevado (normal: 0.1–1.2), mas possível em doença hepática grave.")
print()
print(f"  Direct_Bilirubin: min = {df['Direct_Bilirubin'].min()}, max = {df['Direct_Bilirubin'].max()}")
print(f"    → Valor máximo de 19.7 é elevado (normal: 0.0–0.3), consistente com obstrução biliar.")
print()
print(f"  Alkaline_Phosphotase: min = {df['Alkaline_Phosphotase'].min()}, max = {df['Alkaline_Phosphotase'].max()}")
print(f"    → Valor máximo de 2110 é muito elevado (normal: 44–147), mas possível em doenças hepatobiliares.")
print()
print(f"  Alamine_Aminotransferase (ALT): min = {df['Alamine_Aminotransferase'].min()}, max = {df['Alamine_Aminotransferase'].max()}")
print(f"    → Valor máximo de 2000 é extremamente elevado (normal: 7–56), pode indicar hepatite aguda.")
print()
print(f"  Aspartate_Aminotransferase (AST): min = {df['Aspartate_Aminotransferase'].min()}, max = {df['Aspartate_Aminotransferase'].max()}")
print(f"    → Valor máximo de 4929 é extremamente elevado (normal: 10–40), possível em dano hepático severo.")
print()
print(f"  Total_Protiens: min = {df['Total_Protiens'].min()}, max = {df['Total_Protiens'].max()}")
print(f"    → Intervalo [2.7, 9.6] — plausível (normal: 6.0–8.3).")
print()
print(f"  Albumin: min = {df['Albumin'].min()}, max = {df['Albumin'].max()}")
print(f"    → Intervalo [0.9, 5.5] — plausível (normal: 3.5–5.0). Valores baixos associados a doença hepática.")
print()
print(f"  Albumin_and_Globulin_Ratio: min = {df['Albumin_and_Globulin_Ratio'].min()}, max = {df['Albumin_and_Globulin_Ratio'].max()}")
print(f"    → Intervalo [0.3, 2.8] — plausível (normal: 1.1–2.5).")
print()
print("✓ Nenhuma inconsistência grave detectada. Valores extremos são clinicamente plausíveis.")
print("  Os outliers serão analisados em detalhe na Fase 2 (EDA).")

Verificação de plausibilidade dos valores:
  Age:  min = 4, max = 90
    → Intervalo [4, 90] — plausível, inclui pacientes pediátricos e idosos.

  Total_Bilirubin: min = 0.4, max = 75.0
    → Valor máximo de 75.0 é extremamente elevado (normal: 0.1–1.2), mas possível em doença hepática grave.

  Direct_Bilirubin: min = 0.1, max = 19.7
    → Valor máximo de 19.7 é elevado (normal: 0.0–0.3), consistente com obstrução biliar.

  Alkaline_Phosphotase: min = 63, max = 2110
    → Valor máximo de 2110 é muito elevado (normal: 44–147), mas possível em doenças hepatobiliares.

  Alamine_Aminotransferase (ALT): min = 10, max = 2000
    → Valor máximo de 2000 é extremamente elevado (normal: 7–56), pode indicar hepatite aguda.

  Aspartate_Aminotransferase (AST): min = 10, max = 4929
    → Valor máximo de 4929 é extremamente elevado (normal: 10–40), possível em dano hepático severo.

  Total_Protiens: min = 2.7, max = 9.6
    → Intervalo [2.7, 9.6] — plausível (normal: 6.0–8.3).

  Albumin: min =

## 1.9 — Documentação das Características do Dataset

### Resumo do Dataset

| Característica | Valor |
|---|---|
| **Nome** | Indian Liver Patient Dataset (ILPD) |
| **Fonte** | UCI Machine Learning Repository (via Kaggle) |
| **Registros** | 583 |
| **Features preditoras** | 10 |
| **Variável-alvo** | `Dataset` (1 = doença, 2 = sem doença) |
| **Tipo de problema** | Classificação binária |
| **Classe majoritária** | 1 — doença hepática (416 registros, 71,4%) |
| **Classe minoritária** | 2 — sem doença hepática (167 registros, 28,6%) |
| **Desbalanceamento** | Sim — razão de ~2,49:1 |
| **Variável categórica** | `Gender` (Male: 441, Female: 142) |
| **Valores ausentes** | `Albumin_and_Globulin_Ratio` — 4 registros (0,69%) |
| **Duplicatas** | 13 registros duplicados (26 envolvidos) |
| **Valores negativos** | Nenhum |
| **Inconsistências** | Nenhuma detectada |

### Observações relevantes para as próximas fases

1. **Desbalanceamento das classes:** A classe de doença hepática representa ~71% dos dados. Isso deverá ser considerado nas estratégias de treinamento e na escolha das métricas de avaliação.

2. **Desproporção de gênero:** O dataset contém ~76% de pacientes masculinos e ~24% femininos. Isso pode refletir a composição real da amostra ou um viés de coleta.

3. **Valores extremos:** Algumas features (bilirrubina, enzimas hepáticas) apresentam valores máximos muito distantes da mediana. Esses outliers serão analisados em detalhe na EDA.

4. **Escala das features:** As features possuem escalas muito diferentes (ex.: Age 4–90 vs. Alkaline_Phosphotase 63–2110). Modelos sensíveis à escala precisarão de padronização.

5. **Imputação necessária:** Os 4 valores ausentes em `Albumin_and_Globulin_Ratio` deverão ser imputados, utilizando parâmetros calculados somente nos dados de treinamento.

6. **Transformação do alvo:** A variável `Dataset` deverá ser convertida de {1, 2} para {1, 0} na fase de pré-processamento.

### Checklist de Conclusão da Fase 1

- ✅ Dataset carregado com sucesso sem erros.
- ✅ Todas as colunas identificadas e descritas.
- ✅ Tipos de dados verificados e documentados.
- ✅ Dimensão do dataset confirmada (583 × 11).
- ✅ Distribuição da variável-alvo confirmada e registrada.
- ✅ Valores ausentes identificados e quantificados (4 em `Albumin_and_Globulin_Ratio`).
- ✅ Duplicatas verificadas e decisão documentada (manter por enquanto).
- ✅ Inconsistências verificadas — nenhuma encontrada.
- ✅ Resumo das características do dataset registrado.

In [21]:
import os
os.makedirs('data', exist_ok=True)
df.to_csv('data/df_fase1.csv', index=False)
print('Dataset salvo em data/df_fase1.csv')


Dataset salvo em data/df_fase1.csv
